File: _inc/laravel/config/logging.php
<?php

use App\Config\Constants\SettingsConstants;
use App\Logging\{RemoveTraceProcessor, TabLogs};
use Monolog\Handler\{
    NullHandler,
    StreamHandler,
    SyslogUdpHandler
};

return [
    'default' => env('LOG_CHANNEL', 'stack'),
    'channels' => [
        'stack' => [
            'driver'            => 'stack',
            'channels'          => [
                'volatile', 'short_lived', 'notice', 'warning', 'error', 'critical',
                SettingsConstants::WARN_TRACE, SettingsConstants::ERR_TRACE, SettingsConstants::CRT_TRACE
            ],
            'ignore_exceptions' => false,
        ],

        'volatile' => [
            'driver' => 'daily',
            'path'   => storage_path('logs/volatile.log'),
            'level'  => 'debug',
            'days'   => 2,
            'tap'    => [TabLogs::class]
        ],

        SettingsConstants::WARN_TRACE => [
            'driver' => 'daily',
            'path'   => storage_path('logs/' . SettingsConstants::WARN_TRACE . '.log'),
            'level'  => 'debug',
            'days'   => 2,
            'tap'    => [TabLogs::class]
        ],

        SettingsConstants::ERR_TRACE => [
            'driver' => 'daily',
            'path'   => storage_path('logs/' . SettingsConstants::ERR_TRACE . '.log'),
            'level'  => 'error',
            'days'   => 2,
            'tap'    => [TabLogs::class]
        ],

        'short_lived' => [
            'driver' => 'daily',
            'path'   => storage_path('logs/short_lived.log'),
            'level'  => 'info',
            'days'   => 7,
            'tap'    => [TabLogs::class]
        ],

        'notice' => [
            'driver' => 'daily',
            'path'   => storage_path('logs/notice.log'),
            'level'  => 'notice',
            'days'   => 14,
            'tap' => [
                RemoveTraceProcessor::class,
            ],
        ],

        'warning' => [
            'driver' => 'daily',
            'path'   => storage_path('logs/warning.log'),
            'level'  => 'warning',
            'days'   => 30,
            'tap' => [
                RemoveTraceProcessor::class,
            ],
        ],

        'error' => [
            'driver' => 'daily',
            'path'   => storage_path('logs/error.log'),
            'level'  => 'error',
            'days'   => 60,
            'tap' => [
                RemoveTraceProcessor::class,
            ],
        ],

        'critical' => [
            'driver' => 'single',
            'path'   => storage_path('logs/critical.log'),
            'level'  => 'critical',
            'tap' => [
                RemoveTraceProcessor::class,
            ],
        ],

        SettingsConstants::CRT_TRACE => [
            'driver' => 'daily',
            'path'   => storage_path('logs/' . SettingsConstants::CRT_TRACE . '.log'),
            'level'  => 'critical',
            'tap'    => [TabLogs::class]
        ],

        'slack' => [
            'driver' => 'slack',
            'url' => env('LOG_SLACK_WEBHOOK_URL'),
            'username' => 'Laravel Log',
            'emoji' => ':boom:',
            'level' => env('LOG_LEVEL', 'critical'),
        ],

        'papertrail' => [
            'driver' => 'monolog',
            'level' => env('LOG_LEVEL', 'debug'),
            'handler' => env('LOG_PAPERTRAIL_HANDLER', SyslogUdpHandler::class),
            'handler_with' => [
                'host' => env('PAPERTRAIL_URL'),
                'port' => env('PAPERTRAIL_PORT'),
                'connectionString' => 'tls://' . env('PAPERTRAIL_URL') . ':' . env('PAPERTRAIL_PORT'),
            ],
        ],

        'stderr' => [
            'driver' => 'monolog',
            'level' => env('LOG_LEVEL_STDERR', 'debug'),
            'handler' => StreamHandler::class,
            'formatter' => env('LOG_STDERR_FORMATTER'),
            'with' => [
                'stream' => 'php://stderr',
            ],
        ],

        'syslog' => [
            'driver' => 'syslog',
            'level' => env('LOG_LEVEL_SYS', 'debug'),
        ],

        'errorlog' => [
            'driver' => 'errorlog',
            'level' => env('LOG_LEVEL_ERR', 'debug'),
        ],

        'null' => [
            'driver' => 'monolog',
            'handler' => NullHandler::class,
        ],

        'emergency' => [
            'path' => storage_path('logs/laravel.log'),
        ],

        'PayTabs' => [
            'driver' => 'single',
            'path' => storage_path('logs/paytabs.log'),
            'level' => 'info',
        ],
    ],
    'deprecations' => [
        'channel' => env('LOG_DEPRECATIONS_CHANNEL', 'null'),
        'trace' => false,
    ],
];

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/configs/logging.py
import os
from pathlib import Path

BASE_DIR = Path(__file__).resolve().parent.parent

# Environment values (defaulting as in Laravel)
LOG_CHANNEL = os.environ.get("LOG_CHANNEL", "stack")
LOG_LEVEL = os.environ.get("LOG_LEVEL", "DEBUG").upper()
LOG_SLACK_WEBHOOK_URL = os.environ.get("LOG_SLACK_WEBHOOK_URL", "")
PAPERTRAIL_URL = os.environ.get("PAPERTRAIL_URL", "")
PAPERTRAIL_PORT = os.environ.get("PAPERTRAIL_PORT", "")

# Django logging configuration
LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "verbose": {
            "format": "[%(asctime)s] %(levelname)s [%(name)s:%(lineno)s] %(message)s",
            "datefmt": "%Y-%m-%d %H:%M:%S",
        },
        "simple": {
            "format": "%(levelname)s %(message)s",
        },
    },
    "handlers": {
        # Single file handler (Laravel 'single')
        "file": {
            "level": LOG_LEVEL,
            "class": "logging.FileHandler",
            "filename": os.path.join(BASE_DIR, "storage", "logs", "laravel.log"),
            "formatter": "verbose",
        },
        # Daily rotating file handler (Laravel 'daily')
        "daily": {
            "level": LOG_LEVEL,
            "class": "logging.handlers.TimedRotatingFileHandler",
            "filename": os.path.join(BASE_DIR, "storage", "logs", "laravel.log"),
            "when": "D",       # Rotate daily
            "interval": 1,
            "backupCount": 14, # Keep logs for 14 days
            "formatter": "verbose",
        },
        # Slack handler – requires a custom handler or a third‑party package
        "slack": {
            "level": os.environ.get("LOG_LEVEL", "CRITICAL"),
            "class": "path.to.CustomSlackHandler",  # You need to implement or install a Slack handler.
            "webhook_url": LOG_SLACK_WEBHOOK_URL,
            "username": "Laravel Log",
            "emoji": ":boom:",
        },
        # STDERR handler (similar to Laravel 'stderr')
        "stderr": {
            "level": LOG_LEVEL,
            "class": "logging.StreamHandler",
            "stream": "ext://sys.stderr",
            "formatter": "simple",
        },
        # Syslog handler
        "syslog": {
            "level": LOG_LEVEL,
            "class": "logging.handlers.SysLogHandler",
            # On Linux, typically '/dev/log' is used. Adjust for your OS.
            "address": "/dev/log",
            "formatter": "verbose",
        },
        # Error log using stderr (Laravel 'errorlog')
        "errorlog": {
            "level": LOG_LEVEL,
            "class": "logging.StreamHandler",
            "stream": "ext://sys.stderr",
            "formatter": "simple",
        },
        # Null handler (Laravel 'null')
        "null": {
            "level": LOG_LEVEL,
            "class": "logging.NullHandler",
        },
        # Emergency channel writes to file as a fallback
        "emergency": {
            "level": "ERROR",
            "class": "logging.FileHandler",
            "filename": os.path.join(BASE_DIR, "storage", "logs", "laravel.log"),
            "formatter": "verbose",
        },
        # Custom PayTabs channel
        "PayTabs": {
            "level": "INFO",
            "class": "logging.FileHandler",
            "filename": os.path.join(BASE_DIR, "storage", "logs", "paytabs.log"),
            "formatter": "verbose",
        },
    },
    "loggers": {
        # The root logger will use a "stack" of handlers (here we use "file" as an example).
        "": {
            "handlers": ["file"],
            "level": LOG_LEVEL,
            "propagate": True,
        },
        # You can override Django's default logger if needed.
        "django": {
            "handlers": ["file"],
            "level": LOG_LEVEL,
            "propagate": False,
        },
        # For deprecation warnings you might set:
        "py.warnings": {
            "handlers": ["null"],
            "propagate": False,
        },
    },
}
